server/routes/shared_api/node.py

```diff
@@ -45,3 +62,95 @@ def get_property_value(direction):
     prop = request.json['prop']
   response = fetch.raw_property_values(dcids, prop, direction == 'out')
   return Response(json.dumps(response), 200, mimetype='application/json')
+
+
+def parse_property_expression(expression: str) -> List[PropertySpec]:
+  """
+  Takes a pv expression and parses it into a list of property specs. The
+  expression may chain multiple properties, while each property spec defines a
+  single property to get values for.
+  """
+  expression_matches = re.finditer(_PROPERTY_EXPRESSION_RE, expression)
+  result = []
+  for match in expression_matches:
+    direction, prop, constraints = match.groups()
+    result.append(
+        PropertySpec(prop=prop,
+                     out=direction == _OUT_ARROW,
+                     constraints=constraints or ''))
+  return result
+
+
+def _get_values_for_property_list(dcids: List[str], props: List[PropertySpec]):
+  """
+  Recursive function to get the values for a list of chained properties for a
+  list of dcids. 
+
+  Args:
+    dcids: list of dcids to get the property values for
+    props: list of chained property values
+
+  Returns:
+    map of dcid to final list of values from the chain of props
+  """
+
+  if len(props) == 0:
+    return {}
+
+  # Get the result for the first prop in the prop list and return that result if
+  # that's the only prop in the list.
+  first_prop_result = fetch.raw_property_values(dcids, props[0].prop,
+                                                props[0].out,
+                                                props[0].constraints)
+  if len(props) == 1:
+    return first_prop_result
+
+  # Get the results for the remaining props in the prop list
+  next_dcids = set()
+  for vals in first_prop_result.values():
+    next_dcids.update([v['dcid'] for v in vals if 'dcid' in v])
+  remaining_prop_results = _get_values_for_property_list(
+      sorted(list(next_dcids)), props[1:])
+
+  # Generate results by mapping the results for the remaining props to their
+  # original dcids
+  result = {}
+  for dcid, values in first_prop_result.items():
+    # For each value in this result, get the values for the remaining props
+    result_values = []
+    for val in values:
+      val_dcid = val.get('dcid', '')
+      if val_dcid not in remaining_prop_results:
+        continue
```

Comment:
no need to check since line 125 would return [] anyway?

Reply:
you're right, updated